### PR TITLE
Correctly render legends linked to a map with a theme set

### LINKED ISSUE #27542: Legend to reflect map theme in layouts/ reports
Author Name: **Verity Miles** (Verity Miles)
Original Redmine Issue: [19717](https://issues.qgis.org/issues/19717)

Redmine category:map_composer/printing

---
It would be great to have a map legend reflect a map theme in the same way that the map itself does in a print layout/ report.

I am trying to set up a report in QGIS 3.2 to export a whole series of maps. At the moment, I cannot alter the style of the legend (we have different styles for the one layer depending on the map requirements) without manually changing the style/ map theme in the map viewer window. This effectively means that I have to manually export the series of maps.

Thanks.

---

Related issue(s): #24694 (duplicates)
Redmine related issue(s): [16795](https://issues.qgis.org/issues/16795)

---

### PR DESCRIPTION
**Pull request by nyalldawson**
If a legend is linked to a map which is set to show a particular theme, then correctly follow that same theme's styling when rendering the legend.

Fixes layout legends always render using the canvas' visible theme styling, instead of that of the linked map.

Fixes #27542, #24694, #28919, #28925

Fix sponsored by the Victorian Planning Authority

Peek 2020-04-02 14-23

### CODE REVIEW

**Comment by nirvn** [comment]
Thanks good.

**Comment by DelazJ** [comment]
> Fixes #27542, #24694, #28919, #28925

Thanks Nyall. Afaik each issue report needs its own "fixes" mention, otherwise only the first one (#27452) will be closed when this PR is merged. See eg
image
VS
image

**Comment by nyalldawson** [comment]
@roya0045 that should be working -- the unit test includes a case where a style has a different set of nodes in a theme. But if you want to pull this branch and test, it'd be very welcome!

File: src/core/layout/qgslayoutitemlegend.cpp [hunk 1]
```diff
@@ -821,6 +875,11 @@ void QgsLayoutItemLegend::doUpdateFilterByMap()
   mForceResize = true;
 }
 
+QString QgsLayoutItemLegend::themeName() const
+{
+  return mThemeName;
```
Comment (nirvn) [inline src/core/layout/qgslayoutitemlegend.cpp:880]: @nyalldawson , any reason why this isn't simply in the header file?

**Review by nirvn — APPROVED** [review]

File: src/core/layout/qgslayoutitemlegend.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/layout/qgslayoutitemlegend.cpp:880]: There's no real benefit to it being there -- this isn't performance critical code

Comment (nirvn) [reply src/core/layout/qgslayoutitemlegend.cpp:880]: Okeydokey. Merge away! :)

**Comment by roya0045** [comment]
@nyalldawson In this case, I'll rebase my branch with my PRs in limbo to see of that solves the issue.

**Comment by roya0045** [comment]
OK MXE and my window build is bjorked with the protobuf requirement. I won't be able to test soon.

**Pull request merged**